examples/ethereum-tracker/src/service.rs

```diff
@@ -40,11 +46,134 @@ impl Service for EthereumTrackerService {
 
     async fn handle_query(&self, request: Request) -> Response {
         let schema = Schema::build(
-            self.state.clone(),
+            Query(self.clone()),
             Operation::mutation_root(),
             EmptySubscription,
         )
         .finish();
         schema.execute(request).await
     }
 }
+
+/// The service handler for GraphQL queries.
+pub struct Query(EthereumTrackerService);
+
+#[async_graphql::Object]
+impl Query {
+    /// Returns the Ethereum endpoint the service is using.
+    async fn ethereum_endpoint(&self) -> String {
+        self.0.state.ethereum_endpoint.get().clone()
+    }
+
+    /// Returns the Ethereum contract address that the application is monitoring.
+    async fn contract_address(&self) -> String {
+        self.0.state.contract_address.get().clone()
+    }
+
+    /// Returns the Ethereum block height where monitoring started.
+    async fn start_block(&self) -> u64 {
+        *self.0.state.start_block.get()
+    }
+
+    /// Returns the map of known Ethereum addresses that have used the contract and their balances.
+    async fn accounts(&self) -> &MapView<String, U256Cont> {
+        &self.0.state.accounts
+    }
+
+    /// Reads the initial Ethereum event emitted by the monitored contract.
+    async fn read_initial_event(&self) -> InitialEvent {
+        let start_block = *self.0.state.start_block.get();
+        let events = self
+            .read_events("Initial(address,uint256)", start_block, start_block + 1)
+            .await;
+
+        assert_eq!(events.len(), 1);
+        let event = &events[0];
+        let EthereumDataType::Address(address) = event.values[0].clone() else {
+            panic!("wrong type for the first entry");
+        };
+        let EthereumDataType::Uint256(value) = event.values[1] else {
+            panic!("wrong type for the second entry");
+        };
+
+        InitialEvent {
+            address,
+            balance: value,
+        }
+    }
+
+    /// Reads the transfer events emitted by the monitored Ethereum contract.
+    async fn read_transfer_events(&self, end_block: u64) -> Vec<TransferEvent> {
+        let start_block = *self.0.state.start_block.get();
+        let events = self
+            .read_events(
+                "Transfer(address indexed,address indexed,uint256)",
+                start_block,
+                end_block,
+            )
+            .await;
```

Comment:
What I find a little odd is that now the code has two ways to access the events.
Here, we have the `read_transfer_events` and in `fn update` a different code.

Reply:
But `update` is using `read_transfer_events`. How else could we move this functionality into the service and use it from the contract?